### ISSUE TITLE
New ALI failures in CEE builds

### ISSUE THREAD

**Issue by ikalash** [SNLComputation/Albany#555]
It seems we have some new ALI failures in a couple of the CEE builds: 

http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=92899
http://cdash.sandia.gov/CDash-2-3-0/viewTest.php?onlyfailed&buildid=92900

I think most of them have sensitivities.  The errors look as follows: 

```
Kokkos::DualView::modify_device ERROR: Concurrent modification of host and device views in DualView "exports"
```
@mperego , @bartgol : any idea what happened?  It must be on the Trilinos side...

**Comment by mperego** [comment]
Yeah.. Kokkos and kokkos kernels were snapshot into Trilinos Monday night.
@ndellingwood, do you know what could cause this error:
```Kokkos::DualView::modify_device ERROR: Concurrent modification of host and device views in DualView "exports"```

**Comment by bartgol** [comment]
Yeah, that's suspicious. Also, this is a Host build, so device and host views should be the same view. Something must have changed in either Tpetra or the Thyra adapters, when it comes to mark host/dev views as modified (perhaps during one of the getData1d calls).

**Comment by bartgol** [comment]
> Yeah.. Kokkos and kokkos kernels were snapshot into Trilinos Monday night.

It's hard to believe they did not test Tpetra, but it may be that some ThyraTpetraAdapters features were not fully tested.

**Comment by ikalash** [comment]
We should bring in someone from Kokkos to weigh in.  @ibaned perhaps you have a sense of what changed in Kokkos that could have this effect?

**Comment by mperego** [comment]
Yeah, @ndellingwood pinged above develops Kokkos and he is the one that makes sure that all Trilinos tests pass before snapshotting Kokkos into trilinos.

**Comment by ikalash** [comment]
Ah ok.  I missed your comment about Mauro - sorry about that.

**Comment by ndellingwood** [comment]
@bartgol @mperego @ikalash 

> It's hard to believe they did not test Tpetra, but it may be that some ThyraTpetraAdapters features were not fully tested to adapt to changes in Kokkos/Tpetra.

Yes Tpetra is tested, I used the ATDM scripts to test Cuda builds of Trilinos with all packages enabled on Waterman and Ride, as well as the PR autotester coverage which requires a passing Cuda build in order to be merged. If the functionality that is causing the errors you report has tests in Trilinos, then they may be disabled within the ATDM and/or PR testing scripts, or there is a configuration that would have caught this that I am not aware of. 

Deprecated code is disabled as the default with this kokkos release; there have been changes to the `dual_view` in Kokkos and Tpetra that this affects and were addressed a couple releases ago. Can you point me to where the usage of DualViews occur in the `ThyraTpetraAdapters` as a starting point for me to triage? From the failing CDASH output posted it's not immediately clear to me where the error occurs. I'll do some digging of my own for `modify_device` calls.

Once we chase this down we can determine where the gap in testing coverage is that let this slip through.

In the meantime, disabled code can be reenabled with the macro
`-DKokkos_ENABLE_DEPRECATED_CODE=ON`

**Comment by ndellingwood** [comment]
I've got a host build going, will let you know what I find.

**Comment by bartgol** [comment]
@ndellingwood Mine was just a cold guess about the Thyra adapters, and I would assume that if something is wrong there, other Thyra-dependent packages (like panzer) would have caught it, but it's worth double checking.

You could take a look at thyra/adapters/tpetra/src/Thyra_TpetraMultiVector_def.hpp.

A while ago I had an issue (trilinos/Trilinos#3180) that boiled down to DualView logic not being handled correctly. I am speculating that if Kokkos/Tpetra changed the way they handle sync logic and/or the logic of how/when host/dev are marked as modified, then something may have broken in the Thyra adapters.

**Comment by bartgol** [comment]
@ndellingwood This may also be relevant: did Tpetra/Kokkos change the way it initializes a dualview at construction time, in terms of marking it modified on host/dev?

It looks like one of the tests crashes when it tries to scatter a vector for the first time, and it might be that the src/tgt in the import operation had not been touched since construction time. The relative code you could peek into is in `thyra/adapters/tpetra/src/Thyra_TpetraVectorSpace_def.hpp`, specifically the method `createMember` and `createMembers`: the vecotrs are _not_ zero-ed out. Does this mean that somehow dev/host are not synced, since I'm assuming they are un-inited, and so both marked as modified?

**Comment by ndellingwood** [comment]
@bartgol thank you for the added info. I haven't reproduced in Trilinos and will give you an update what I tried so far and some comments regarding the failure messages and how they may relate to the information you've provided. Could I also get a Trilinos configuration used in this build?

Based on the output of the failing Albany test, I took a guess at a few packages to enable in hopes of getting lucky with a relatively fast Trilinos build. No luck, but here's the configuration I used:

Trilinos develop SHA aa11e22afc6c864ca63b4564637b8638d87b1ee7

configure
```bash

module purge
module load sems-env
module load sems-cmake/3.12.2 sems-gcc/7.2.0 sems-git sems-openmpi/4.0.2 sems-ninja_fortran/1.8.2 sems-superlu/4.3
#sems-hdf5/1.8.12/parallel sems-netcdf/4.4.1/exo_parallel sems-zlib/1.2.8/base sems-boost/1.59.0/base 

export TRILINOS_PATH=${PWD}/../..

INSTALL_LOCATION=${PWD}/install
 
cmake \
-D CMAKE_INSTALL_PREFIX:PATH="${INSTALL_LOCATION}" \
-D TPL_ENABLE_MPI:BOOL=ON \
-D CMAKE_CXX_COMPILER="`which mpicxx`" \
-D CMAKE_C_COMPILER="`which mpicc`" \
-D CMAKE_Fortran_COMPILER="`which mpifort`" \
-D KOKKOS_ARCH="SKX" \
  -D TPL_ENABLE_BLAS:STRING=ON \
  -D TPL_ENABLE_LAPACK:STRING=ON \
-D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
-D Trilinos_ENABLE_TESTS:BOOL=OFF \
-D Trilinos_ENABLE_EXAMPLES:BOOL=OFF \
\
  -DCMAKE_BUILD_TYPE=DEBUG \
  -DTrilinos_ENABLE_TESTS=OFF \
  -DTPL_ENABLE_SUPERLU=ON \
  -DTrilinos_ENABLE_Kokkos=ON \
    -DKokkos_ENABLE_SERIAL=ON \
    -DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=ON \
  -DTrilinos_ENABLE_Tpetra=ON \
    -DTpetra_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Amesos2=ON \
  -DTrilinos_ENABLE_MueLu=ON \
    -DMueLu_ENABLE_TESTS=ON \
  -DTrilinos_ENABLE_Thyra=ON \
    -DThyra_ENABLE_TESTS=ON \
${TRILINOS_PATH}

```

All tests passed, which include Tpetra, Thyra, and Thyra Tpetra adapters.

 Configure option in the Albany build of Trilinos must be enabling Kokkos debugging (or subset of the options) that must result in setting the configure option 

`-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=ON `

(either explicitly or by setting Kokkos_ENABLE_DEBUG which could active this).

That configure option allows for DualViews to check the modify flags in the `modify*` routines and error out if the host and device views have been concurrently modified, without a synchronization in between. For example, marking a host view modified, updating its contents, and then requesting to make the device view as modified and attempt to modify it, without first synchronizing the device view to the (modified) host view prior to attempting to modify the device view data.
In a host build, the state of the "host" and "device" views are still tracked separately and can result in this error happening, as the pattern should still be followed regardless.


Regarding this comment:

> we need to make sure that every time one grabs nonconst stuff from a thyra vector/mv, the underlying tpetra/kokkos dualview is marked as modified

my suspicion is prior to marking the dualview modified, there needs to be a check if synchronization is necessary (and if so to synchronize),

```c++
//dv symbol for a DualView
if (dv.need_sync_device()) {
  dv.sync_device();
}
//continue on to mark the dual as modified in anticipation of upcoming updates to the data

When the nonconst stuff is grabbed from thrya, does this happen in Albany or in a Trilinos package? If in Trilinos can you point me to some specific routines where this is facilitated? 

I'll enabled Panzer and test (I hadn't seen your message in time before I kicked off first Trilinos builds) and see if this tickles the potential Thyra adapter culprit with useful info.

In the meantime, if this is a critical blocker for Albany, I suggest adding this configure option to the Trilinos builds for the time being:
`-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF`

**Comment by bartgol** [comment]
@ndellingwood Mmm, given that your ThyraTpetraAdapters tests passed, I'm leaning towards the "it's our fault" diagnosis.

Could you please comment on my last post? Basically, are un-initialized Tpetra (multi)vectors a potential source of these errors? Do we need to explicitly initialize them before we do an import/export operation?

**Comment by ndellingwood** [comment]
@bartgol 

> are un-initialized Tpetra (multi)vectors a potential source of these errors? Do we need to explicitly initialize them before we do an import/export operation?

I'm not certain on the Tpetra side if this is a source of error, and would have to look at the code in more detail (a Tpetra developer may be able to answer more quickly), in particular I need to understand how the MVs (which use dual views) are constructed without initialization. 
With dual views, it is possible to construct a dual view from uninitialized views which will be marked as having the same "sync" state after construction, and could cause issues if not carefully dealt with (note the user or caller is responsible for ensuring the data is synchronized when using this type of ctor for proper use), for example:

```c++
      using scalar_type = double;
      using memory_space = Kokkos::HostSpace;
      using dual_view_t = Kokkos::DualView<scalar_type*, memory_space>;
      using view_t = Kokkos::View<scalar_type*, memory_space>;

      // to create an uninitialized dual_view, the device and host views need to be created uninitialized first then passed to the dual view ctor - note constructing in this way requires the user ensure they're synchronized, so in this example could be problematic
      view_t dv(Kokkos::ViewAllocateWithoutInitializing("dv"), size);
      auto hv = Kokkos::create_mirror_view(dv);
      dual_view_t duv3(dv,hv);
      std::cout << "need_sync_host: " << duv3.need_sync_host() << std::endl;
      std::cout << "need_sync_device: " << duv3.need_sync_device() << std::endl;
      // modify flags are identical, and so the dual view is considered "synchronized"

      // Note that the call below is not supported for dual views
      //dual_view_t duv4(Kokkos::ViewAllocateWithoutInitializing("duv4"), size);
```

**Comment by bartgol** [comment]
@ndellingwood I think the relevant piece of code for those mv construction is this (from Tpetra_MultiVector_def.hpp):

```
      using Kokkos::view_alloc;
      using Kokkos::WithoutInitializing;
     ...
     d_view = dev_view_type (view_alloc (label, WithoutInitializing), lclNumRows, numCols);
     ....
     dual_view_type dv (d_view, Kokkos::create_mirror_view (d_view));
     // Whether or not the user cares about the initial contents of the
     // MultiVector, the device and host views are out of sync.  We
     // prefer to work in device memory.  The way to ensure this
     // happens is to mark the device view as modified.
     dv.modify_device ();
     return dvl
```

If this is how the dual view is created, could there be issues? It looks like the device view is marked modified, and the comment there seems to be related to the fact that the views are out of sync. Now, what would happen if this multivector was fed as the destination vector for an import/export operation? I'm guessing that there could be an issue. Do you have a unit test of tpetra that is quick to hack, and do an import/export operation (with combine mode ADD) into a (multi)vector that was created with zeroOut=false? I suspect (hope?) that the test would throw. If that's so, then I think the burden is on our side, to makce sure that vectors are not used in import/export ops if they have not been explicitly initialized.

**Comment by bartgol** [comment]
Actually, I'm going to try to hack the ImportExport_UnitTests.cpp in Tpetra, to verify my conjecture. I'll post here the result.

**Comment by ikalash** [comment]
Thanks for looking into this @ndellingwood .  If you'd like to try to reproduce this, the configure options used in the nightlies for Trilinos can be found here: they are the union of https://github.com/SNLComputation/Albany/blob/master/doc/dashboards/cee-compute011.sandia.gov/ctest_nightly.cmake#L441-L573 and https://github.com/SNLComputation/Albany/blob/master/doc/dashboards/cee-compute011.sandia.gov/ctest_nightly.cmake#L593-L643 .  The tests seem to fail in our CEE gcc and clang debug builds.  The modules for these are https://github.com/SNLComputation/Albany/blob/master/doc/dashboards/cee-compute011.sandia.gov/modules_debug.sh and https://github.com/SNLComputation/Albany/blob/master/doc/dashboards/cee-compute011.sandia.gov/modules_clang.sh.  It sounds like Luca is working on this on the Albany side, so perhaps this information is no longer needed.  Luca, should I hold off adding the configure option ``-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF`` @ndellingwood suggested for the time building so that we can see from the nightlies if any changes you make fix the problem?

**Comment by bartgol** [comment]
No. The fact that either src or dst were not initialized does not seem to affect import/export operations.

**Comment by ndellingwood** [comment]
@bartgol Just finished up a couple meetings, I spoke with Karen about the vector ctor and she shared essentially what you posted above, that there is an option to set `zeroOut=false` which hits the code path you have posted, where an uninitialized device view is created, then its host mirror, and then the device view is marked modified (so the flags are out of sync at this point).

The question seems to be how the usage pattern for an uninitialized MV differs from that of the default case when the views are initialized to zeros. Like you mention, does this mean import/export ops require an explicitly initialized MV, or if the MV is uninitialized, is this allowed with but required that users ensure the views are synchronized (i.e. same modify flags) before import/export operations?

Adding @kddevin and @mhoemmen from the Tpetra team - can you help clarify the question about uninitialized MV usage from @bartgol, in particular do import/export ops require an explicitly initialized MV? If uninitialized, is the requirement on the user to synchronize the views before the import/export op?

**Comment by ndellingwood** [comment]
> No. The fact that either src or dst were not initialized does not seem to affect import/export operations.

@bartgol just to check, you had `-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=ON` in your configure?

**Comment by ndellingwood** [comment]
@bartgol I have this reproduced in Trilinos, will work on it and keep you updated.

**Comment by mperego** [comment]
@ndellingwood  that's great! Thanks for working on it!

**Comment by ndellingwood** [comment]
So far I've chased down where the `modify_device` call happens in Tpetra that triggers the DualView error, but it's not clear to me if it is due to some misuse elsewhere in Trilinos.

Error is triggered in `tpetra/core/src/Tpetra_MultiVector_def.hpp` in the `packAndPrepare` routine after a check if a realloc is necessary by `reallocDualViewIfNeeded` - in the error-triggering case `packOnHost` is false, resulting in a call to `exports.modify_device ()`, however in the error case the host flags had already been marked as modified (without a sync call that would clear the flags) and this results in the error
`Kokkos::DualView::modify_device ERROR: Concurrent modification of host and device views in DualView "exports"` when `-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=ON` is set.

**Comment by ndellingwood** [comment]
Follow on to previous message:

There is a comment prior to setting the `exports` modify flags (host vs device is determined earlier, the flags are being set in anticipation of computation later in the routine), the comment mentions that the `exports` flags will be cleared after resizing, but the `reallocDualViewIfNeeded` only calls resizes if the sizes differ else it simply returns without doing anything besides a size comparison leaving the flags unchanged (that is, the flags are not reset).

If I make changes in the `packAndPrepare` routine in `tpetra/core/src/Tpetra_MultiVector_def.hpp` to check if `exports` needs to be synced and do so if needed prior to setting the flags, the failing muelu tests pass (I'd been using `MueLu_UnitTestsTpetra_MPI_4` for testing), though I'm uncertain this is the correct way to handle this (tpetra unit tests are still passing in my configuration).

```c++
    if (packOnHost) {
      if (exports.need_sync_host ())
        exports.sync_host ();
      exports.modify_host ();
    }      
    else {
      if (exports.need_sync_device ())
        exports.sync_device ();
      exports.modify_device ();
    }
```

@mhoemmen @kddevin what are your thoughts on this scenario? 
To summarize, there is a case where the `exports` host flags are modified and `packOnHost==false`, a resize is unnecessary and so the modify flags are never reset, but this then results in a dual view error when setting `exports.modify_device()` in anticipation of computation later in the routine.
Is the code snip above appropriate to handle this scenario, or is there some sort of misuse that happened prior resulting in this scenario (which would require a different strategy to catch, though I'm not sure what that should be)?

**Comment by ndellingwood** [comment]
@bartgol @mperego @ikalash I put in a PR to Trilinos with a potential fix, trilinos/Trilinos#6768

**Comment by ndellingwood** [comment]
@bartgol @mperego @ikalash PR  trilinos/Trilinos#6768 just merged into Trilinos develop branch, hopefully this resolves the problems you are seeing.

**Comment by ikalash** [comment]
Seems this has fixed most of the test failures, with the exception of the RigidBody tests: http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=5368064&build=92956 .  Any thoughts on why that test might still be having the issue given the Trilinos PR that was merged in?

**Comment by ndellingwood** [comment]
@ikalash I just put in a PR trilinos/Trilinos#6782 to fix an occurrence of this issue in Panzer, hopefully that is the final fix needed for the test failures.

**Comment by mperego** [comment]
@ndellingwood I doubt that PR trilinos/Trilinos#6782 will fix the issue. We do not use Panzer Dof Manager in Albany. Anyway thanks for looking into this so quickly.

**Comment by bartgol** [comment]
@ndellingwood Thanks for taking the time to track it down!

**Comment by bartgol** [comment]
@ikalash I think the RigidBody test may have other issues. The initial condition passed to NOX has zero-residual, which is very suspicious at best. I'm not excluding the presence of a DV sync bug still, somewhere in Albany or Trilinos, but it may be worth finding out why f=0 at iter 0 first; might be the crash will disappear when that's fixed.

**Comment by bartgol** [comment]
Btw, I think we do have an issue in dual view handling, specifically when it comes to using  `getLocalDiagCopy` from Tpetra_CrsMatrix. I think we need to init the vector to 0, or, better, to use the trick used in the fix by @ndellingwood to _ignore_ modified flags, since getLocalDiagCopy overwrites everything anyways (the issue is that it starts by marking it modified on host before launching the kernel).

I am however getting worried about some other stuff (see #trilinos/Trilinos6791).

I will push a fix to getDiagCopy straight to master. This should fix the crash, but I suspect won't fix the test (for the reason I explained in the previous comment).

**Comment by ikalash** [comment]
It seems we're getting NaNs now in an Aeras test:

Aeras_TC5_explicitHV_nu8e15_vort_BackwardEuler 

http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=5381032&build=93063 

```
Aeras::SpectralDiscretization::writeSolution: writing time 0 to index 1 in file explicit_BE_nu8e15_vort.exo
 Time = 0
  final time = 0
          Response[0] = 1.22464680e-15  2.00000000e+01  -0.00000000e+00 -0.00000000e+00 0.00000000e+00  2.01992580e+267 1.27142235e+11  3.68816527e+08  0.00000000e+00  1.27142770e+11  
   
   Entering Rythmos::BackwardEulerStepper<double>::takeStep(1.20000000e+03,STEP_TYPE_FIXED) ...

p=0: *** Caught standard std::exception of type 'Thyra::CatastrophicSolveFailure' :

 /home/ikalash/nightlyAlbanyTests/Results/Trilinos/packages/stratimikos/adapters/belos/src/Thyra_BelosLinearOpWithSolve_def.hpp:660:
 
 Throw number = 2
 
 Throw test that evaluated to true: true
 
 /home/ikalash/nightlyAlbanyTests/Results/Trilinos/packages/belos/src/BelosStatusTestImpResNorm.hpp:633:
 
 Throw number = 1
```
@bartgol: could this be due to your changes?

**Comment by bartgol** [comment]
I doubt it.

**Comment by ikalash** [comment]
How about some of these ```zeroOut=true``` options discussed above?

**Comment by bartgol** [comment]
The way vectors are init-ed is still the same: no zero-out by default on the ThyraTpetraAdapters side. Nothing in terms of vector content should have changed, I think. The only difference is how we handled the DV syncronization, but that should not generate nans where there were none before.

**Comment by mhoemmen** [comment]
In debug mode with `zeroOut=true`, Tpetra::MultiVector prefills with NaNs to help debugging.

**Comment by bartgol** [comment]
@mhoemmen so in debug mode Tpetra does not actually zeroOut the vector? Or is it that it does two passes, one filling with NaNs, and one filling with 0's? Also, are NaNs FPE safe (meaning quiet_NaN)?

**Comment by ikalash** [comment]
@mhoemmen : thanks for the tip - I was guessing something like this might have started happening.  One somewhat puzzling this is the NaNs show up in non-debug builds...

**Comment by mhoemmen** [comment]
@bartgol wrote:
> so in debug mode Tpetra does not actually zeroOut the vector? Or is it that it does two passes, one filling with NaNs, and one filling with 0's?

`zeroOut=true` _always_ means "fill the entries with zeros."  In debug mode, `zeroOut=false` means ["fill the entries with quiet NaNs"](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp#L203).

**Comment by mhoemmen** [comment]
@ikalash wrote:
> I was guessing something like this might have started happening.

Tpetra has been doing that for years, though.

**Comment by ikalash** [comment]
I don't get why we get NaNs now w/ a non-debug build...

**Comment by ikalash** [comment]
The lingering issue with the one Aeras test has been moved to issue #558.
